app/ui-react/packages/api/src/helpers/integrationFunctions.ts

```diff
@@ -200,6 +193,454 @@ export function getStepIcon(apiUri: string, step: Step): string {
+  propertyDefinitionStep: ActionDescriptorStep
+) {
+  return getDefaultsFromProperties(propertyDefinitionStep.properties!);
+}
+
+/**
+ * Extracts out the default configuredProperty values for the supplied propertyDefinitionSteps
+ * @param propertyDefinitionSteps
+ */
+function getPropertyDefaults(
+  propertyDefinitionSteps: ActionDescriptorStep[] = []
+) {
+  return propertyDefinitionSteps
+    .map(getDefaultsForPropertyDefinitionStep)
+    .reduce((result, current) => {
+      return { ...result, ...current };
+    }, {});
+}
+
+/**
+ * Inspects the supplied data shape and determines if the user set it on the step or not
+ * @param dataShape
+ */
+export function isUserDefinedDataShape(dataShape?: DataShape) {
+  return (
+    dataShape && dataShape.metadata && dataShape.metadata.userDefined === 'true'
+  );
+}
+
+/**
+ * Checks if the supplied step has either an input or output data shape
+ * @param step
+ * @param isInput
+ */
+export function hasDataShape(step: Step, isInput = false) {
+  if (!step.action || !step.action.descriptor) {
+    return false;
+  }
+  const descriptor = step.action.descriptor;
+  const dataShape = isInput
+    ? descriptor.inputDataShape
+    : descriptor.outputDataShape;
+  return (
+    dataShape &&
+    dataShape.kind &&
+    dataShape.kind.toLowerCase() !== DataShapeKinds.NONE
+  );
+}
+
+/**
+ * Returns whether or not the supplied descriptor has an input or output datashape of ANY
+ * @param descriptor
+ */
+export function isActionShapeless(descriptor: ActionDescriptor) {
+  if (!descriptor) {
+    return false;
+  }
+  const inputDataShape = descriptor.inputDataShape;
+  const outputDataShape = descriptor.outputDataShape;
+  return (
+    inputDataShape &&
+    outputDataShape &&
+    (inputDataShape.kind && outputDataShape.kind) &&
+    (inputDataShape.kind.toLowerCase() === DataShapeKinds.ANY ||
+      outputDataShape.kind.toLowerCase() === DataShapeKinds.ANY)
+  );
+}
+
+/**
+ * Returns whether or not the supplied descriptor has an input datashape of ANY
+ * @param descriptor
+ */
+export function isActionInputShapeless(descriptor: ActionDescriptor) {
+  if (!descriptor) {
+    return false;
+  }
+  const inputDataShape = descriptor.inputDataShape;
+  return (
+    inputDataShape &&
+    inputDataShape.kind &&
+    inputDataShape.kind.toLowerCase() === DataShapeKinds.ANY
+  );
+}
+
+/**
+ * Returns whether or not the supplied descriptor has an output datashape of ANY
+ * @param descriptor
+ */
+export function isActionOutputShapeless(descriptor: ActionDescriptor) {
+  if (!descriptor) {
+    return false;
+  }
+  const outputDataShape = descriptor.outputDataShape;
+  return (
+    outputDataShape &&
+    outputDataShape.kind &&
+    outputDataShape.kind.toLowerCase() === DataShapeKinds.ANY
+  );
+}
+
+/**
+ * Filters connections based on the supplied position in the step array
+ * @param steps
+ * @param position
+ */
+export function filterStepsByPosition(
```

Comment:
We actually already have this imported here -> https://github.com/syndesisio/syndesis-react/blob/master/app/ui-react/syndesis/src/modules/integrations/components/editor/utils.ts#L251 and it's being used already.

Reply:
but it's cool, I'll pull this one out next PR :-)